File: python/프로그래머스/미로탈출명령어.py
def solution(n, m, x, y, r, c, k):
    moves = [(1, 0), (0, -1), (0, 1), (-1, 0)]  # d, l, r, u
    moves_str = ['d', 'l', 'r', 'u']

    # 맨해튼 거리 계산 함수
    def cal_dist(sx, sy, ex, ey):
        return abs(sx - ex) + abs(sy - ey)

    # 현재 위치에서 목표 위치까지 이동이 가능한지 판단
    dist = cal_dist(x, y, r, c)
    if dist > k or (k - dist) % 2 == 1:
        return "impossible"

    # Greedy 방식으로 경로 탐색
    result = []
    while k > 0:
        for i in range(4):
            nx, ny = x + moves[i][0], y + moves[i][1]

            # 유효한 좌표인지 확인
            if 1 <= nx <= n and 1 <= ny <= m:
                # 이동 후 남은 거리가 유효한지 확인
                remaining_dist = cal_dist(nx, ny, r, c)
                if remaining_dist <= k - 1 and (k - 1 - remaining_dist) % 2 == 0:
                    result.append(moves_str[i])
                    x, y = nx, ny
                    k -= 1
                    break

    return ''.join(result)

File: java/미로탈출명령어.java
import java.util.*;

class Solution {
    int N, M, R, C, K;
    String answer = null;
    // 사전식 순서: d, l, r, u
    int[] dx = {1, 0, 0, -1};
    int[] dy = {0, -1, 1, 0};
    char[] mvs = {'d', 'l', 'r', 'u'};

    public String solution(int n, int m, int x, int y, int r, int c, int k) {
        N = n; M = m; R = r - 1; C = c - 1; K = k;

        int dist = Math.abs((x - 1) - R) + Math.abs((y - 1) - C);
        if (dist > k || (k - dist) % 2 != 0) return "impossible";

        dfs(x - 1, y - 1, new StringBuilder(), 0);

        return answer == null ? "impossible" : answer;
    }

    void dfs(int cx, int cy, StringBuilder sb, int count) {
        if (answer != null) return;

        int remainingDist = Math.abs(cx - R) + Math.abs(cy - C);
        if (remainingDist > K - count) return;

        if (count == K) {
            if (cx == R && cy == C) {
                answer = sb.toString();
            }
            return;
        }

        for (int i = 0; i < 4; i++) {
            int nx = cx + dx[i];
            int ny = cy + dy[i];

            if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                sb.append(mvs[i]);
                dfs(nx, ny, sb, count + 1);
                sb.deleteCharAt(sb.length() - 1);
            }
        }
    }
}
